Building-OSX.md

```diff
@@ -1,2 +1,171 @@
+
+We use Homebrew to install the libraries on which MLDB depends.  MLDB comes packaged with some libraries
+(those for which we need to build specially or a specific version is required); the rest of them are either
+automatically available from the system, or installed here.  (We aim to make the list smaller over time and
+in particular remove the dependency on `boost` eventually).
+
+```
+brew install gnu-time icu4c libmagic boost boost-python3 coreutils libssh2 lz4 openssl python@3.9 snappy v8 xz yaml-cpp libb2
+```
+
+If your're on the Mac M1, some of the Python dependencies also require a Rust compiler to build:
+
+```
+brew install rust
+```
+
+We also need Python's `virtualenv` command, and a couple of other system-wide Python packages
+that can't easily be installed into a virtualenv for various reasons:
+
+```
+pip3 install virtualenv python-dateutil bottle requests
+```
+
+## Check out MLDB
+
+We first get MLDB from it's Git repository.  There are two ways to do it, depending upon if your machine is
+authenticated via its SSH key with GitHub (if you're not sure, choose the first).
+
+### 1: If you're not on GitHub or your machine isn't authenticated via its SSH key (anyone)
+
+In this case, you can build MLDB but you can't push changes back upstream.
+
+```
+git clone https://github.com/mldbai/mldb.git
+```
+
+### 2: If you're on GitHub and can access it via SSH (developers):
+
+In this case, you have access to the full set of GitHub functionality and can push code
+back to GitHub.
+
+```
+git clone git@github.com:mldbai/mldb.git
+```
+
+## Change into the MLDB subdirectory
+
+The rest of the build instructions require you to have done this step and it's easy to
+forget, so don't!
+
+```
+cd mldb
+```
+
+## Pull the submodules
+
+These are large: the data used for testing MLDB is about 100MB big, so it may take a while.
+We ask `git` to do four submodules at a time in parallel in order to speed it up a bit.
+
+```
+git submodule update --init --recursive --jobs=4
+```
+
+## Install the compilation cache (optional)
+
+If you plan on doing lots of MLDB development, compilation times can be greatly reduced using a
+compilation cache.  For OSX, we use sccache.
+
+```
+curl -fsSL https://gitlab.com/mldbai/mldb_build/-/raw/main/sccache-`uname -m`-darwin.gz | gzip -d > ./sccache && chmod +x ./sccache
+declare -x COMPILER_CACHE=./sccache
+```
+
+## Install MLDB's Python dependencies
+
+These are installed inside MLDB's virtual environment, not system-wide, and so won't interfere
+with the rest of the Python stuff that's going on in your system.
+
+```
+make dependencies
+```
+
+## Compile MLDB
+
+This process is long.  You should use an argument to `-j` which is equal to the number of processors cores
+in your system; be aware however that some files require lots of memory to build (more than 500MB) and
+so you also need to ensure that your machine has sufficient RAM for the compilation.
+
+```
+make -j6 -k compile
```

Comment:
we used to do `nice make -j $(nproc) compile` :)

Reply:
That unfortunately runs my laptop out of memory (it thinks it has 12 cores, but there are only 6 real ones, and I don't have enough memory to compile simultaneously on the 12).